app/src/test/java/me/jorgecastillo/hiroaki/MockForeverTests.kt

```diff
@@ -0,0 +1,122 @@
+import java.io.IOException
+
+class MockForeverTests : MockServerSuite() {
+
+    private lateinit var dataSource: GsonNewsNetworkDataSource
+
+    @Before
+    override fun setup() {
+        super.setup()
+        dataSource = GsonNewsNetworkDataSource(
+            server.retrofitService(GsonConverterFactory.create())
+        )
+    }
+
+    @Test
+    fun shouldRespondMultipleTimes() {
+        server.whenever(Method.GET, "v2/top-headlines")
+            .thenRespondForever(success(jsonBody = fileBody("GetNews.json")))
+
+        repeat(3) {
+            runBlocking { dataSource.getNews() }
+        }
+
+        server.verify("v2/top-headlines").called(times(count = 3))
+    }
+
+    @Test
+    fun shouldDispatchMultipleTimes() {
+        server.whenever(Method.GET, "v2/top-headlines")
+            .thenDispatchForever { request ->
+                success(
+                    jsonBody = inlineBody(
+                        """{
+                              "status": "ok",
+                              "totalResults": 2342,
+                              "articles": [
+                                {
+                                  "source": {
+                                    "id": ${request.path.length},
+                                    "name": "Lifehacker.com"
+                                  },
+                                  "author": "Jacob Kleinman",
+                                  "title": "How to Get Android P's Screenshot Editing Tool on Any Android Phone",
+                                  "description": "Last year, Apple brought advanced screenshot editing tools to the iPhone with iOS 11, and, this week, Google fired back with a similar Android feature called Markup. The only catch is that this new tool is limited to Android P, which launches later this year …",
+                                  "url": "https://lifehacker.com/how-to-get-android-ps-screenshot-editing-tool-on-any-an-1823646122",
+                                  "urlToImage": "https://i.kinja-img.com/gawker-media/image/upload/s--Y-5X_NcT--/c_fill,fl_progressive,g_center,h_450,q_80,w_800/nxmwbkwzoc1z1tmak7s4.jpg",
+                                  "publishedAt": "2018-03-09T20:30:00Z"
+                                }
+                              ]
+                            }
+                        """
+                    )
+                )
+            }
+
+        repeat(3) {
+            runBlocking { dataSource.getNews() }
+        }
+
+        server.verify("v2/top-headlines").called(times(count = 3))
+    }
+
+    @Test(expected = IOException::class)
+    fun regularResponseShouldOverrideRespondForever() {
+        server.whenever(Method.GET, "v2/top-headlines")
+            .thenRespond(error())
+            .thenRespondForever(success(jsonBody = fileBody("GetNews.json")))
+
+        runBlocking {
+            dataSource.getNews()
+        }
+
+        server.verify("v2/top-headlines").called()
```

Comment:
This test should verify that error is returned on first call, then success with the provided body for subsequent calls. We're just verifying that the endpoint is called here 🤔

Reply:
There is a single call here. So no error at all. I want to demonstrate the behavior where calling it The 2nd times overrides the existing. So error is ignored.

Reply:
As I used I would expect this to return error on a first match, then subsequent calls return the provider success forever 🤔 that's what "then" wording leads to understand and I'd like to achieve.

Reply:
Fair point. Could it be fine to come up with different naming?

Calling forever 2nd time should override the first. This makes it fundamentally different than the original behavior. Let me think about it more and how this can be made obvious. 

If you can think of any alternative, it would be awesome.

Reply:
I think the naming is great and I like the feature, I'd aim for matching the behaviour I describe. That would make it coherent, as in living up to the "then" prefix for ordering matched / dispatched requests. That's what makes the whole thing very idiomatic on usage, imho.

My expectations for this feature would be:
* If you add any number of `thenRespond` before `thenRespondForever`, those must be matched first and removed (or flagged as already matched, depending on implementation changes required - probably required for the following points).
* As soon as a `thenRespondForever` is declared, further `thenRespond` calls declared after it **would have no effect**. This would be clarified in docs, since `thenRespondForever` has a wider priority given to its own semantics. It's "forever".
* Same for multiple calls to `thenRespondForever`, I'd suggest keeping the **first one** so it matches previous behavior also.
* We can add an additional `resetQueue` param to `thenRespond` and `thenRespondForever` (same for the dispatch variants) so you can reset the queue and your newly added matcher becomes the first one in the queue. That would be used for cases where you want to assert multiple scenarios under the same test. In my opinion it's always better to write different tests for this, so they're atomic in terms of error reporting once they start failing. Still it's fine to provide this feature for flexibility imho, since every project has different requirements and people uses to work in many different ways, so I'd rather not being super opinionated on this. Some scenario for the sake of the example:

  - User configures matchers for one request, one `thenRespond()`, one `thenRespondForever()`.
  - `thenRespond()` gets matched when first endpoint call occurs, user verifies.
  - 3 more endpoint calls occur, `thenRespondForever()` is matched for all those. User verifies.
  - Then user decides she/he wants to queue a new `thenRespond()` (or a new `thenRespondForever()`) and make further calls, then verify. The user could use the `resetQueue` param for this. `resetQueue` param would be defaulted to false.

Thoughts? Do you see any flaws? 🤔

Reply:
Sorry for super late reply. I think I will give it another go.

I think this is going to be very difficult because technically, it is impossible to compare matchers.

Imagine a scenario where `thenRespond` and `thenRespondForever` has slightly different matchers matching the same request. This also depends on the request, that's why all should be recorded and then decided at the time of the actual request. 

Queue logic is much easier since they are processed and removed sequentially.

If you already have an idea on the data structure, please let me know. I think I will first right the tests according to these expectations. And then see how we can do this.